### DISCUSSION TITLE
Performing the attack on a GPU device?

### DISCUSSION THREAD

**Discussion by kfarivar** [Trusted-AI/adversarial-robustness-toolbox#1385] [Q&A] [answered]
Hello
I wanted to know if it is possible to run attacks on a gpu ?

As far as I looked into the code the generate methods for attacks only accept np.ndarray as inputs and there is no parameter when initializing an attack to set a GPU device. 
any help is appreciated.

**Comment by beat-buesser** [comment, marked as answer]
Hi @kfarivar Thank you very much for using ART! Yes, of course you can use GPUs. Setting a specific device depends on the deep learning framework you are using. Which framework are you using?

**Reply by kfarivar** [reply]
I'm using Pytorch.

**Reply by beat-buesser** [reply]
For PyTorch, for example `PyTorchClassifier` accepts an argument `device_type` that can be either `gpu` or `cpu`. For `gpu` it will select the current device.

**Reply by awskkzl** [reply]
Hi @beat-buesser, is there any way to make the attack towards SklearnClassifier be running on gpu?

**Reply by beat-buesser** [reply]
Hi @awskkzl I think the answer is no for `SklearnClassifier`, not because of the attack or ART but because it seems that sklearn is not supporting GPUs in the near future: https://scikit-learn.org/stable/faq.html#will-you-add-gpu-support

**Reply by le000043** [reply]
Hello @beat-buesser, is GPU the default device on `TensorFlowV2Classifier`? Thanks

**Reply by beat-buesser** [reply]
Hi @le000043 With TensorFlow you can set specific devices outside of ART following https://www.tensorflow.org/guide/gpu. Without any specific settings TensorFlow will automatically run on a device if one is available.

**Comment by UsamaShakoor** [comment]
hello,
i am unable to compute GraphiteBlackBox attack on gpu.
**My implementation:**
from art.estimators.classification import PyTorchClassifier
from art.attacks.evasion import GRAPHITEBlackbox

classifier = PyTorchClassifier(model=model, device_type="cuda",loss=nn.CrossEntropyLoss(), input_shape= ([10,3,112, 112]), nb_classes=8)

noise_size = (112,112)
net_size = (112,112)
attack = GRAPHITEBlackbox(classifier=classifier, noise_size=noise_size, net_size=net_size,num_xforms_mask=50, num_xforms_boost=50, num_boost_queries=500)

image_adv = attack.generate(x=x_test_class_0, y=y_test_class_0, x_tar=x_tar_replicated)

**Where**
x=x_test_class_0, y=y_test_class_0, x_tar=x_tar_replicated are all pytorch tensors

**which generates following error:**
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
[<ipython-input-34-4b376607807f>](https://localhost:8080/#) in <cell line: 18>()
     16 
     17 
---> 18 image_adv = attack.generate(x=x_test_class_0, y=y_test_class_0, x_tar=x_tar_replicated)

2 frames
[/usr/local/lib/python3.10/dist-packages/art/attacks/evasion/graphite/graphite_blackbox.py](https://localhost:8080/#) in generate(self, x, y, **kwargs)
    241             clip_min, clip_max = self.estimator.clip_values
    242         else:
--> 243             clip_min, clip_max = np.min(x), np.max(x)
    244 
    245         # target initialization image

[/usr/local/lib/python3.10/dist-packages/numpy/core/fromnumeric.py](https://localhost:8080/#) in min(a, axis, out, keepdims, initial, where)
   2951     6
   2952     """
-> 2953     return _wrapreduction(a, np.minimum, 'min', axis, None, out,
   2954                           keepdims=keepdims, initial=initial, where=where)
   2955 

[/usr/local/lib/python3.10/dist-packages/numpy/core/fromnumeric.py](https://localhost:8080/#) in _wrapreduction(obj, ufunc, method, axis, dtype, out, **kwargs)
     84                 return reduction(axis=axis, dtype=dtype, out=out, **passkwargs)
     85             else:
---> 86                 return reduction(axis=axis, out=out, **passkwargs)
     87 
     88     return ufunc.reduce(obj, axis, dtype, out, **passkwargs)

TypeError: min() received an invalid combination of arguments - got (out=NoneType, axis=NoneType, ), but expected one of:
 * ()
 * (Tensor other)
 * (int dim, bool keepdim)
      didn't match because some of the keywords were incorrect: out, axis
 * (name dim, bool keepdim)
      didn't match because some of the keywords were incorrect: out, axis

Any help is appreciated.

**Reply by beat-buesser** [reply]
Hi @UsamaShakoor To run on GPUs, try setting `device_type="gpu"` in `PyTorchClassifier`.

**Reply by UsamaShakoor** [reply]
Hi, @beat-buesser 
Even with `device_type="gpu"`, I'm still encountering the same error. However, if I convert `x=x_test_class_0`, `y=y_test_class_0`, and `x_tar=x_tar_replicated` to numpy arrays, then this error disappears. But in that case, it doesn't run on the GPU. 
I want to speedup the process

**Reply by beat-buesser** [reply]
I took a closer look and it looks like `GRAPHITEBlackbox` is not implemented with support for GPUs. Are you able to reproduce the original paper?
